### ISSUE TITLE
False negative: override of `@final` property

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#9795]
The typing spec indicates that `@final` should be enforced for properties, but pyright doesn't emit an error in this case.

```python
from typing import final


class Base:
    @final
    @property
    def prop(self) -> int:
        return 1


class Child(Base):
    # This should result in an error.
    @property
    def prop(self) -> int:
        return 2
```

Refer to [this thread](https://github.com/python/typing/pull/1916).

**Comment by jimmyb579** [comment]
from abc import ABCMeta

class FinalPropertyMeta(ABCMeta):
    def __new__(mcs, name, bases, namespace):
        cls = super().__new__(mcs, name, bases, namespace)
        if '__final_properties__' not in cls.__dict__:
            cls.__final_properties__ = set()
        
        # Collect all final properties from current class definition
        cls.__final_properties__.update(
            attr_name 
            for attr_name in namespace 
            if isinstance(getattr(cls.__dict__[attr_name], 'fget', None), property)
               and getattr(namespace[attr_name].fget._is_final_property_, False)
        )
        
        # Check if any base class has defined these properties as final
        for base_class in bases:
            if hasattr(base_class.__dict__.get('__metaclass__', {}), '__bases__'):
                base_final_properties = getattr(base_class.__dict__, '__final_properties__', set())
                intersections = base_final_properties.intersection(cls.__dict__)
                if intersections:
                    raise TypeError(f"Cannot override final properties {intersections} in {name}")

        return cls


def _is_final_property(func):
    func._is_final_property_ = True
    return func


class FinalProperty(metaclass=FinalPropertyMeta):
    pass


class Base(FinalProperty):
    @property @_is_final_property_
    def prop(self) -> int:
        return 1


class Child(Base):    
    # This will raise a TypeError at runtime due to metaclass check.
    @property 
    def prop(self) -> int:  
        return 2

**Comment by erictraut** [comment]
The spec is currently unclear on whether the `@final` applies to the attribute (which is a descriptor object) or the method (which is wrapped by the descriptor object). This matters because if it applies to the attribute, then it's not possible to define a setter or a deleter on the property. If it applies to the method, then it's not clear what it means for override behaviors if, for example, the getter is marked `@final` but the setter is not. I'd prefer to have these details ironed out in the spec before I implement this because the implementation will differ significantly between these different interpretations.

**Comment by jimmyb579** [comment]
Thank you. I'm not proceeding at this time on this project.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.394.
